### PR TITLE
chore(sequencer)!: update storage keys locations and values (ENG-898)

### LINKED ISSUE #1611: Update storage keys in sequencer
It is difficult to be sure that a given storage key is unique since they are defined in multiple different locations in the sequencer codebase, and we have no convention regarding naming/prefixes.

For each component, move its keys (const values and functions generating DB keys) to a new module named `keys` inside the component's `storage` module.  All keys should be given a prefix relevant to the component's name.

┆Issue Number: ENG-898

### PR DESCRIPTION
**Pull request by Fraser999**
## Summary
The storage keys for all components were moved from the top of  `state_ext` modules into new `storage::keys` modules.  The key values were updated to a consistent style.

## Background
We wanted the keys to be easy to find and to follow a consistent pattern in terms of their values and locations.

## Changes
- All key consts and functions were moved to `<COMPONENT>/storage/keys`.
- All keys were given a prefix matching the component name followed by a `/`.
- All keys now use full words (except for a few instances of well-know abbreviations like "init") and `_` as word separators.
- Existing helper structs in various `state_ext` modules for formatting keys were unified into a new struct `AddressPrefixer` in `crate::storage`.
- The `Asset` functionality was moved from `crate::storage_keys` to `crate::storage`.

## Testing
- All keys have snapshot tests (including const keys - this might be considered overkill?)
- All keys have tests ensuring they have a component name as prefix.
- All prefix consts or functions which are used in `state_ext` modules are tested to ensure they are actually prefixes of the given keys.

## Breaking Changelist
- This breaks the on-disk format of both the verifiable and non-verifiable store.

## Related Issues
Closes #1611.

### CODE REVIEW

File: crates/astria-sequencer/src/assets/storage/keys.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
+use std::borrow::Cow;
+
+use astria_core::primitive::v1::asset::IbcPrefixed;
+use astria_eyre::eyre::{
+    bail,
+    ContextCompat,
+    Result,
+    WrapErr,
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/assets/storage/keys.rs:8]:
```suggestion
    WrapErr as _,
```

File: crates/astria-sequencer/src/assets/storage/keys.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
+use std::borrow::Cow;
+
+use astria_core::primitive::v1::asset::IbcPrefixed;
+use astria_eyre::eyre::{
+    bail,
+    ContextCompat,
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/assets/storage/keys.rs:6]: I think `ContextCompat` and `WrapErr` are essentially the same as of eyre 0.6.12. Remove this one?

File: crates/astria-sequencer/src/assets/storage/keys.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    Result,
+    WrapErr,
+};
+
+use crate::storage::keys::Asset;
+
+pub(in crate::assets) const NATIVE_ASSET_KEY: &str = "assets/native_asset";
+pub(in crate::assets) const BLOCK_FEES_PREFIX: &[u8] = b"assets/block_fees/";
+pub(in crate::assets) const FEE_ASSET_PREFIX: &[u8] = b"assets/fee_asset/";
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/assets/storage/keys.rs:15]: For consistency, replace by a str?

File: crates/astria-sequencer/src/assets/state_ext.rs [hunk 4]
```diff
@@ -20,46 +25,23 @@ use tendermint::abci::{
...
-    )
-}
-
 /// Creates `abci::Event` of kind `tx.fees` for sequencer fee reporting
-fn construct_tx_fee_event<T: std::fmt::Display>(
-    asset: &T,
-    fee_amount: u128,
-    action_type: String,
-) -> Event {
+fn construct_tx_fee_event<T: Display>(asset: &T, fee_amount: u128, action_type: String) -> Event {
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/assets/state_ext.rs:44]: While we are at it, can we could make this `fn construct_tx_fee_event<T: Display, U: Protobuf>(asset: &T, fee_amount: u128) -> Event {` and then in the function body use `<U as Protobuf>::full_name`

File: crates/astria-sequencer/src/app/storage/keys.rs [hunk 5]
```diff
@@ -0,0 +1,35 @@
+pub(in crate::app) const CHAIN_ID_KEY: &str = "app/chain_id";
+pub(in crate::app) const REVISION_NUMBER_KEY: &str = "app/revision_number";
+pub(in crate::app) const BLOCK_HEIGHT_KEY: &str = "app/block_height";
+pub(in crate::app) const BLOCK_TIMESTAMP_KEY: &str = "app/block_timestamp";
+
+pub(in crate::app) fn storage_version_by_height_key(height: u64) -> Vec<u8> {
+    format!("app/storage_version/{height}").into_bytes()
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/app/storage/keys.rs:7]: As for our other comments: IMO this should be a string. (also note: here `height` is using its display impl which I agree with, whereas elsewhere we convert with `to_le_bytes`, which is inconsistent).

**Review by SuperFluffy — COMMENTED** [review]
This is great, thank you for going through all the keys and streamlining them.

I very much like that we have full (I think?) coverage of all keys and prefixes.

Since this is a breaking change and a rework of how our fees are structured, I think we should break with our previous key and make their types and (where applicable) byte-encoding consistent everywhere.

Some open points that I think should be addressed before merge:

1. make the type of all keys consistently `&str`/`String`. Right now we have a string representation in some places, and raw byte slices in others. It's not clear why that's done. Yet especially for snapshot tests it's very valuable to have a human readable representation.
2. Apart from ibc-prefixed assets (which are `"ibc/<hex-encoded-32-bytes>"`, all key/prefix parts that involve raw bytes (like account) should be base64 encoded to save space.
3. I don't have a strong opinion on only using the byte part of ibc-prefixed assets when constructing keys from them. I mildly lean toward encoding the full string (including `"ibc/"`) and not just the bytes.

File: crates/astria-sequencer/src/grpc/storage/keys.rs
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/grpc/storage/keys.rs:1]: > It's confusing that the keys in some modules are String, while these are bare Vec<u8>. Let's just be consistent about this everywhere (and resort to bytes)?

The cnidarium crate requires `String` keys for verifiable storage and `Vec<u8>` keys for no-verifiable storage, so I think it makes sense to have the keys defined in the type to suit their specific use cases.

> I think this file might be a good use case of defining the component prefix as a constant and using it everywhere, for example via one of the [const_format macros](https://docs.rs/const_format/0.2.33/const_format/index.html)

I'm not sure I see a benefit of that over just using `format!` here.  (As discussed offline, I'll change the contents of the `Vec<u8>` keys to be human-readable.

File: crates/astria-sequencer/src/bridge/storage/keys.rs
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/bridge/storage/keys.rs:1]: > Can we make all of the keys strings (for human readability, consistency across all storage keys)?

As above, I'd rather keep the types as is to suit the use-cases.

> There are so many keys in here, I think we should define a constant and use const_format to create the (stringy) keys and prefixes.

As before, I don't see a benefit over using `format!`.

File: crates/astria-sequencer/src/authority/storage/keys.rs [hunk 6]
```diff
@@ -0,0 +1,26 @@
+pub(in crate::authority) const SUDO_KEY: &str = "authority/sudo";
+pub(in crate::authority) const VALIDATOR_SET_KEY: &str = "authority/validator_set";
+pub(in crate::authority) const VALIDATOR_UPDATES_KEY: &[u8] = b"authority/validator_updates";
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/authority/storage/keys.rs:3]: Agreed offline to make all keys human-readable, but I'd prefer to keep the types as is.

File: crates/astria-core/src/primitive/v1/asset/denom.rs [hunk 7]
```diff
@@ -543,20 +565,22 @@ pub struct IbcPrefixed {
 }
 
 impl IbcPrefixed {
+    pub const LENGTH: usize = 32;
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-core/src/primitive/v1/asset/denom.rs]: Done in [6dff230](https://github.com/astriaorg/astria/pull/1616/commits/6dff230972661693c60a62426e814886acc4b7bb).

File: crates/astria-sequencer/src/storage/keys.rs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    },
+    str::FromStr,
+};
+
+use astria_core::primitive::v1::asset::IbcPrefixed;
+
+use crate::accounts::AddressBytes;
+
+/// Helper struct whose `Display` impl outputs the prefix followed by the hex-encoded address.
+pub(crate) struct AddressPrefixer<'a, T> {
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/storage/keys.rs]: Done in [6dff230](https://github.com/astriaorg/astria/pull/1616/commits/6dff230972661693c60a62426e814886acc4b7bb).

File: crates/astria-sequencer/src/ibc/state_ext.rs [hunk 9]
```diff
@@ -24,42 +27,20 @@ use tracing::{
     instrument,
 };
 
-use super::storage;
+use super::storage::{
+    self,
+    keys::{
+        channel_balance_key,
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/ibc/state_ext.rs]: Done in [6dff230](https://github.com/astriaorg/astria/pull/1616/commits/6dff230972661693c60a62426e814886acc4b7bb).

File: crates/astria-sequencer/src/bridge/state_ext.rs [hunk 10]
```diff
@@ -28,111 +28,28 @@ use tracing::{
     instrument,
 };
 
-use super::storage;
+use super::storage::{
+    self,
+    keys::{
+        asset_id_key,
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/bridge/state_ext.rs]: Done in [6dff230](https://github.com/astriaorg/astria/pull/1616/commits/6dff230972661693c60a62426e814886acc4b7bb).

File: crates/astria-sequencer/src/ibc/storage/keys.rs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        AddressPrefixer,
+        Asset,
+    },
+};
+
+pub(in crate::ibc) const IBC_SUDO_KEY: &str = "ibc/sudo";
+pub(in crate::ibc) const ICS20_WITHDRAWAL_BASE_FEE_KEY: &str = "ibc/ics20_withdrawal_base_fee";
+const IBC_RELAYER_PREFIX: &str = "ibc/relayer/";
+
+/// Example: `ibc/channel-xxx/balance/0101....0101`.
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/ibc/storage/keys.rs]: Done in [9966331](https://github.com/astriaorg/astria/pull/1616/commits/9966331ef78c4b77b5c1929d72b6de2978d558d1).

File: crates/astria-sequencer/src/grpc/state_ext.rs [hunk 12]
```diff
@@ -22,33 +22,19 @@ use cnidarium::{
 };
 use tracing::instrument;
 
-use super::storage;
+use super::storage::{
+    self,
+    keys::{
+        block_hash_by_height_key,
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/grpc/state_ext.rs]: Done in [6dff230](https://github.com/astriaorg/astria/pull/1616/commits/6dff230972661693c60a62426e814886acc4b7bb).

File: crates/astria-sequencer/src/assets/state_ext.rs (hunk 4 above)
Comment (Fraser999) [reply crates/astria-sequencer/src/assets/state_ext.rs]: Done in [bb0c659](https://github.com/astriaorg/astria/pull/1616/commits/bb0c65935434d38b7028291c2ce4b774bfe7d78b).

File: crates/astria-sequencer/src/grpc/storage/keys.rs
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/grpc/storage/keys.rs:1]: Changed to use strings in [11bafd5](https://github.com/astriaorg/astria/pull/1616/commits/11bafd5b4882c901267fed4909cfcadeeb669504).

File: crates/astria-sequencer/src/grpc/storage/keys.rs [hunk 13]
```diff
@@ -0,0 +1,63 @@
+use astria_core::primitive::v1::RollupId;
+
+pub(in crate::grpc) fn block_hash_by_height_key(height: u64) -> Vec<u8> {
+    [b"grpc/block_hash/".as_slice(), &height.to_le_bytes()].concat()
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/grpc/storage/keys.rs]: Done in [11bafd5](https://github.com/astriaorg/astria/pull/1616/commits/11bafd5b4882c901267fed4909cfcadeeb669504).

File: crates/astria-sequencer/src/grpc/storage/keys.rs [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+pub(in crate::grpc) fn sequencer_block_header_by_hash_key(hash: &[u8; 32]) -> Vec<u8> {
+    [b"grpc/block_header/", hash.as_slice()].concat()
+}
+
+pub(in crate::grpc) fn rollup_data_by_hash_and_rollup_id_key(
+    hash: &[u8; 32],
+    rollup_id: &RollupId,
+) -> Vec<u8> {
+    [b"grpc/rollup_data/", hash.as_slice(), rollup_id.as_ref()].concat()
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/grpc/storage/keys.rs]: Done in [11bafd5](https://github.com/astriaorg/astria/pull/1616/commits/11bafd5b4882c901267fed4909cfcadeeb669504).

File: crates/astria-sequencer/src/grpc/storage/keys.rs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+pub(in crate::grpc) fn rollup_transactions_proof_by_hash_key(hash: &[u8; 32]) -> Vec<u8> {
+    [b"grpc/rollup_txs_proof/", hash.as_slice()].concat()
+}
+
+pub(in crate::grpc) fn rollup_ids_proof_by_hash_key(hash: &[u8; 32]) -> Vec<u8> {
+    [b"grpc/rollup_ids_proof/", hash.as_slice()].concat()
+}
+
+#[cfg(test)]
+mod tests {
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/grpc/storage/keys.rs:45]: Done in [11bafd5](https://github.com/astriaorg/astria/pull/1616/commits/11bafd5b4882c901267fed4909cfcadeeb669504).

File: crates/astria-sequencer/src/bridge/storage/keys.rs
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/bridge/storage/keys.rs:1]: Done in [11bafd5](https://github.com/astriaorg/astria/pull/1616/commits/11bafd5b4882c901267fed4909cfcadeeb669504) (other than the usage of `const_format`).

File: crates/astria-sequencer/src/authority/storage/keys.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Fraser999) [reply crates/astria-sequencer/src/authority/storage/keys.rs]: Changed to use strings in [11bafd5](https://github.com/astriaorg/astria/pull/1616/commits/11bafd5b4882c901267fed4909cfcadeeb669504).

File: crates/astria-sequencer/src/assets/storage/keys.rs (hunk 3 above)
Comment (Fraser999) [reply crates/astria-sequencer/src/assets/storage/keys.rs]: Changed to use strings in [11bafd5](https://github.com/astriaorg/astria/pull/1616/commits/11bafd5b4882c901267fed4909cfcadeeb669504).

File: crates/astria-sequencer/src/app/storage/keys.rs (hunk 5 above)
Comment (Fraser999) [reply crates/astria-sequencer/src/app/storage/keys.rs]: Changed to use strings in [11bafd5](https://github.com/astriaorg/astria/pull/1616/commits/11bafd5b4882c901267fed4909cfcadeeb669504).

File: crates/astria-sequencer/src/assets/storage/keys.rs (hunk 1 above)
Comment (Fraser999) [reply crates/astria-sequencer/src/assets/storage/keys.rs]: Done in [1c8f7d1](https://github.com/astriaorg/astria/pull/1616/commits/1c8f7d1ed51e245cb547187ae21510214e3cd490).

File: crates/astria-sequencer/src/assets/storage/keys.rs (hunk 2 above)
Comment (Fraser999) [reply crates/astria-sequencer/src/assets/storage/keys.rs]: Fixed in [1c8f7d1](https://github.com/astriaorg/astria/pull/1616/commits/1c8f7d1ed51e245cb547187ae21510214e3cd490).

File: crates/astria-sequencer/src/assets/state_ext.rs [hunk 16]
```diff
@@ -36,13 +39,13 @@ use super::storage::{
 use crate::storage::StoredValue;
 
 /// Creates `abci::Event` of kind `tx.fees` for sequencer fee reporting
-fn construct_tx_fee_event<T: Display>(asset: &T, fee_amount: u128, action_type: String) -> Event {
+fn construct_tx_fee_event<P: Protobuf, T: Display>(asset: &T, fee_amount: u128) -> Event {
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/assets/state_ext.rs:42]: This is nice.

File: crates/astria-sequencer/src/grpc/storage/keys.rs [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+pub(in crate::grpc) fn rollup_data_by_hash_and_rollup_id(
+    hash: &[u8; 32],
+    rollup_id: &RollupId,
+) -> String {
+    format!(
+        "grpc/rollup_data/{}/{}",
+        Base64Display::new(hash, &URL_SAFE),
+        Base64Display::new(rollup_id.as_ref(), &URL_SAFE)
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/grpc/storage/keys.rs:22]: We should make the rollup ID display impl URL safe in general.

We have a chance to still change the formatting of rollup IDs - should we just do this in this PR?

**Review by SuperFluffy — APPROVED** [review]
This looks great, thank you.

I wonder if in this PR we should just change base64 encoding to the url safe alphabet everywhere (so that we don't have a different encoding in the storage keys and the normal display impl).

File: crates/astria-sequencer/src/grpc/storage/keys.rs (hunk 17 above)
Comment (Fraser999) [reply crates/astria-sequencer/src/grpc/storage/keys.rs]: Done in [4db1136](https://github.com/astriaorg/astria/pull/1616/commits/4db1136280c76390790d920d2f3db9dec6b3ab10).

**Comment by Fraser999** [comment]
> I wonder if in this PR we should just change base64 encoding to the url safe alphabet everywhere (so that we don't have a different encoding in the storage keys and the normal display impl).

Ticketed for follow-up at #1649.

**Pull request merged**